authorization/service/deleteDevices.go

```diff
@@ -0,0 +1,135 @@
+	it := tx.RetrieveByOwner(owner)
+	defer it.Close()
+	if it.Err() != nil {
+		return nil, fmt.Errorf("failed to obtain user devices: %w", it.Err())
+	}
+	var deviceIds []string
+	var d persistence.AuthorizedDevice
+	for it.Next(&d) {
+		deviceIds = append(deviceIds, d.DeviceID)
+	}
+	return deviceIds, nil
+}
+
+func (s *Service) publishDevicesUnregistered(ctx context.Context, owner string, deviceIDs []string) error {
+	v := events.Event{
+		Type: &events.Event_DevicesUnregistered{
+			DevicesUnregistered: &events.DevicesUnregistered{
+				Owner:     owner,
+				DeviceIds: deviceIDs,
+				AuditContext: &events.AuditContext{
+					UserId: owner,
+				},
+				Timestamp: pkgTime.UnixNano(time.Now()),
+			},
+		},
+	}
+	data, err := utils.Marshal(&v)
+	if err != nil {
+		return err
+	}
+
+	err = s.publisher.PublishData(ctx, events.GetDevicesUnregisteredSubject(owner), data)
+	if err != nil {
+		return err
+	}
+
+	err = s.publisher.Flush(ctx)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// DeleteDevices removes a devices from user.
+func (s *Service) DeleteDevices(ctx context.Context, request *pb.DeleteDevicesRequest) (*pb.DeleteDevicesResponse, error) {
+	tx := s.persistence.NewTransaction(ctx)
+	defer tx.Close()
+
+	owner := request.UserId
+	if owner == "" {
+		if token, err := grpc_auth.AuthFromMD(ctx, "bearer"); err == nil {
+			uid, err := grpc.ParseOwnerFromJwtToken(s.ownerClaim, token)
+			if err == nil {
+				owner = uid
+			}
+		}
+	}
+
+	if owner == "" {
+		return nil, log.LogAndReturnError(status.Errorf(codes.InvalidArgument, "cannot delete devices: invalid UserId"))
+	}
+
+	// TODO validate jwt token -> only jwt token is supported
+
+	var deviceIds []string
+	if len(request.DeviceIds) == 0 {
+		var err error
+		if deviceIds, err = getUserDevices(tx, owner); err != nil {
+			return nil, log.LogAndReturnError(status.Errorf(codes.InvalidArgument, "cannot delete devices: %v", err))
+		}
+		if len(deviceIds) == 0 {
+			return &pb.DeleteDevicesResponse{}, nil
+		}
+	} else {
+		deviceIds = getUniqueDeviceIds(request.DeviceIds)
+		if len(deviceIds) == 0 {
+			return nil, log.LogAndReturnError(status.Errorf(codes.InvalidArgument, "cannot delete devices: invalid DeviceIds"))
+		}
+	}
+
+	var deletedDeviceIds []string
+	for _, deviceId := range deviceIds {
+		_, ok, err := tx.Retrieve(deviceId, owner)
+		if err != nil {
+			return nil, log.LogAndReturnError(status.Errorf(codes.Internal, "cannot delete device('%v'): %v", deviceId, err.Error()))
+		}
+		if !ok {
+			log.Debugf("cannot retrieve device('%v') by user('%v')", deviceId, owner)
+			continue
+		}
+
+		err = tx.Delete(deviceId, owner)
+		if err != nil {
+			return nil, log.LogAndReturnError(status.Errorf(codes.NotFound, "cannot delete device('%v'): not found", deviceId))
+		}
+
+		deletedDeviceIds = append(deletedDeviceIds, deviceId)
+	}
+
+	if err := s.publishDevicesUnregistered(ctx, owner, deletedDeviceIds); err != nil {
```

Comment:
I am not really sure if we can send this event. At the moment we introduce sharing of devices, or customers will implement it in their own auth service, it's impossible to distinguish if the device was really deleted or if sharing for user B was stopped, but it's still in the system for user A. Plan was to send this event if device sharing is stopped for user B by user A for example. But then, coap gateway listening on these events cannot decide if it should disconnect it or not. Does this explanation make sense?
@Danielius1922 @jkralik